client/src/session.rs

```diff
@@ -0,0 +1,434 @@
+        session_id: Vec<u8>,
+        timeout: Duration,
+    ) -> anyhow::Result<Dispatched<T>>
+    where
+        proto::response::Kind: TryInto<T, Error = ()>,
+        T: 'static,
+    {
+        let response = self.response::<T>(request.request_id, timeout).await;
+        let packet = proto::Packet {
+            session_id,
+            kind: Some(proto::packet::Kind::Request(request)),
+        };
+        self.send(packet).await?;
+
+        Ok(response.await?)
+    }
+
+    #[inline(always)]
+    pub(crate) async fn response<'a, T>(
+        &self,
+        request_id: RequestId,
+        timeout: Duration,
+    ) -> LocalBoxFuture<'a, anyhow::Result<Dispatched<T>>>
+    where
+        proto::response::Kind: TryInto<T, Error = ()>,
+        T: 'static,
+    {
+        self.dispatcher.clone().response::<T>(request_id, timeout)
+    }
+
+    pub async fn send(&self, packet: impl Into<codec::PacketKind>) -> anyhow::Result<()> {
+        let mut sink = self.sink.clone();
+        Ok(sink.send((packet.into(), self.remote)).await?)
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct StartingSessions {
+    state: Arc<Mutex<StartingSessionsState>>,
+
+    layer: SessionsLayer,
+    sink: OutStream,
+}
+
+struct StartingSessionsState {
+    /// Initialization of session started by other peers.
+    incoming_sessions: HashMap<SessionId, mpsc::Sender<(RequestId, proto::request::Session)>>,
+    /// Temporary sessions stored during initialization period.
+    /// After session is established, it is moved to SessionsLayer.
+    tmp_sessions: HashMap<SocketAddr, Arc<Session>>,
+}
+
+impl StartingSessions {
+    pub fn new(layer: SessionsLayer, sink: OutStream) -> StartingSessions {
+        StartingSessions {
+            state: Arc::new(Mutex::new(StartingSessionsState {
+                tmp_sessions: Default::default(),
+                incoming_sessions: Default::default(),
+            })),
+            sink,
+            layer,
+        }
+    }
+
+    pub fn temporary_session(&mut self, addr: SocketAddr) -> Arc<Session> {
+        let session = Session::new(addr, SessionId::generate(), self.sink.clone());
+        self.state
+            .lock()
+            .unwrap()
+            .tmp_sessions
+            .insert(addr, session.clone());
+        session
+    }
+
+    pub fn remove_temporary_session(&mut self, addr: &SocketAddr) {
+        self.state.lock().unwrap().tmp_sessions.remove(addr);
+    }
+
+    pub fn dispatcher(&self, addr: SocketAddr) -> Option<Dispatcher> {
+        self.state
+            .lock()
+            .unwrap()
+            .tmp_sessions
+            .get(&addr)
+            .map(|session| session.dispatcher.clone())
+    }
+
+    pub async fn dispatch_session(
+        self,
+        session_id: Vec<u8>,
+        request_id: RequestId,
+        from: SocketAddr,
+        request: proto::request::Session,
+    ) {
+        // This will be new session.
+        match session_id.is_empty() {
```

Comment:
After starting writing in rust, I stopped liking `if ... else..`. It's esthetic thing, so I don't change.

Reply:
> also feel it would be cleaner as an option, so its not accidentally creating a new session on a wrong input field
And I don't understand this. It's following the protocol. Empty session means attempt to establish session.